TFEngine.py

```diff
@@ -1857,86 +1857,127 @@ def search(self, dataset, do_eval=True, output_layer_name="output", output_file=
       max_seq_length=max_seq_length,
       used_data_keys=self.network.used_data_keys)
 
-    output_layer = self.network.layers[output_layer_name]
-    out_beam_size = output_layer.output.beam_size
-    output_layer_beam_scores = None
-    if out_beam_size is None:
-      print("Given output %r is after decision (no beam)." % output_layer, file=log.v1)
-    else:
-      print("Given output %r has beam size %i." % (output_layer, out_beam_size), file=log.v1)
-      output_layer_beam_scores = output_layer.get_search_choices().beam_scores
-    target_key = output_layer.target or self.network.extern_data.default_target
+    if not isinstance(output_layer_names, list):
+      output_layer_names = [output_layer_names]
+    num_targets = len(output_layer_names)
+
+    output_layers, out_beam_sizes, output_layer_beam_scores, target_keys = [], [], [], []
+    for output_layer_name in output_layer_names:
+      output_layer = self.network.layers[output_layer_name]
+      output_layers.append(output_layer)
+      out_beam_size = output_layer.output.beam_size
+      if out_beam_size is None:
+        print("Given output %r is after decision (no beam)." % output_layer, file=log.v1)
+        output_layer_beam_scores.append(None)
+      else:
+        print("Given output %r has beam size %i." % (output_layer, out_beam_size), file=log.v1)
+        output_layer_beam_scores.append(output_layer.get_search_choices().beam_scores)
+      out_beam_sizes.append(output_layer.output.beam_size)
+      target_keys.append(output_layer.target or self.network.extern_data.default_target)
 
     out_cache = None
     seq_idx_to_tag = {}
     if output_file:
       assert output_file_format in {"txt", "py"}
-      assert dataset.can_serialize_data(target_key)
+      num_serializable_targets = sum(dataset.can_serialize_data(target_key) for target_key in target_keys)
+      assert num_serializable_targets > 0
       assert not os.path.exists(output_file)
       print("Will write outputs to: %s" % output_file, file=log.v2)
       output_file = open(output_file, "w")
-      out_cache = {}  # corpus-seq-idx -> str|list[(float,str)]
+      out_cache = {}  # corpus-seq-idx -> str|list[(float,str)]|dict[str -> str|list[(float,str)]] depending on number of output layers and beam sizes
```

Comment:
It should not depend on number of output layers.

Reply:
Also not on beam sizes. It actually depends if the output is after a choice (i.e. single best) or before (the whole beam).